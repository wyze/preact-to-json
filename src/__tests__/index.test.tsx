import { Component, h } from 'preact'
import render from '..'

const element = {
  $$typeof: Symbol.for('react.test.json'),
  children: null,
  key: null,
  props: {},
}

describe('preact-to-json', () => {
  it('works with one component', () => {
    const Foo = () => (<div />)
    const actual = render(<Foo />)

    expect(actual).toEqual({ ...element, type: 'div' })
  })

  it('only shallow renders', () => {
    const Child = () => (<div />)
    const Parent = () => (<Child />)
    const Grandparent = () => (<Parent />)
    const actual = render(<Grandparent />)

    expect(actual).toEqual({
      ...element,
      type: 'Parent',
    })
    expect(actual).toMatchSnapshot()
  })

  it('works with attributes', () => {
    const Link = ({ to, text }: any) => (<a href={to}>{text}</a>)
    const actual = render(<Link to="//google.com" text="Google" />)

    expect(actual).toEqual({
      ...element,
      children: [ 'Google' ],
      props: { href: '//google.com' },
      type: 'a',
    })
    expect(actual).toMatchSnapshot()
  })

  it('prints objects nicely', () => {
    const Child = ( props: any ) => (<div />)
    const Parent = ( props: any ) => (<Child {...props} />)
    const actual = render(<Parent style={{ color: 'rebeccapurple', outline: 0 }} />)

    expect(actual).toEqual({
      ...element,
      props: { style: { color: 'rebeccapurple', outline: 0 } },
      type: 'Child',
    })
    expect(actual).toMatchSnapshot()
  })

  it('supports displaying functions', () => {
    const Child = ( props: any ) => (<div />)
    const Parent = ( props: any ) => (<Child {...props} />)
    const actual = render(<Parent handleClick={jest.fn()} />)

    expect(actual).toEqual({
      ...element,
      props: { handleClick: expect.any(Function) },
      type: 'Child',
    })
    expect(actual).toMatchSnapshot()
  })

  it('makes pretty snapshots', () => {
    const Header = ({ text }: any) => (<header>{text}</header>)
    const Item = ({ text }: any) => (<li>{text}</li>)
    const List = ({ items }: any) => (<ul>{items.map(( item: any ) => <Item text={item} />)}</ul>)
    const Container = () => (<div><Header text="List" /><List items={[ 1, 2, 3, 4, 5 ]} /></div>)
    const actual = render(<Container />)

    expect(actual).toEqual({
      ...element,
      children: [
        { ...element, type: 'Header', props: { text: 'List' } },
        { ...element, type: 'List', props: { items: [ 1, 2, 3, 4, 5 ] } },
      ],
      type: 'div',
    })
    expect(actual).toMatchSnapshot()
  })

  it('works with class components', () => {
    class Foo extends Component<{}, {}> {
      public render() {
        return <div />
      }
    }
    const actual = render(<Foo />)

    expect(actual).toEqual({ ...element, type: 'div' })
    expect(actual).toMatchSnapshot()
  })
})
