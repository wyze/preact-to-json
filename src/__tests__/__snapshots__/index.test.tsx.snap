// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preact-to-json makes pretty snapshots 1`] = `
<div>
  <Header
    text="List"
  />
  <List
    items={
      Array [
        1,
        2,
        3,
        4,
        5,
      ]
    }
  />
</div>
`;

exports[`preact-to-json only shallow renders 1`] = `<Parent />`;

exports[`preact-to-json prints objects nicely 1`] = `
<Child
  style={
    Object {
      "color": "rebeccapurple",
      "outline": 0,
    }
  }
/>
`;

exports[`preact-to-json supports displaying functions 1`] = `
<Child
  handleClick={[Function]}
/>
`;

exports[`preact-to-json works when component returns null 1`] = `undefined`;

exports[`preact-to-json works with attributes 1`] = `
<a
  href="//google.com"
>
  Google
</a>
`;

exports[`preact-to-json works with class components 1`] = `<div />`;

exports[`preact-to-json works with class components who render null 1`] = `undefined`;
